Implement role wise access to project summaries

Sujanadh opened the pull request:
## What type of PR is this? (check all applicable)

- [X] 🍕 Feature
- [ ] 🐛 Bug Fix
- [ ] 📝 Documentation
- [ ] 🧑‍💻 Refactor
- [ ] ✅ Test
- [ ] 🤖 Build or CI
- [ ] ❓ Other (please specify)

## Related Issue

- Issue: #2311
- osm login issue in mapper page

## Describe this PR

- Create a `public_view` dependency to allow public users without login to view the project summaries. The main goal of introducing it is to get the user information if they are logged in so that we can check if the user has access to the private projects or not. If `login_required` is used here user won't be able to view project cards without logging in.

- Check if the user is `superadmin`; if he is `superadmin`, he can view every project.

- Check if the user is `org_admin`; then he can view all the projects within the organisation, whether the projects are private or public, along with the other public projects from other organisations.

- Check if the user is a mapper and he has access to private projects; otherwise, he will only view public projects.

- Update osm callback endpoint in mapper page `callback/osm -> callback/osm/mapper`
## TODO

- Implement other project visibility categories such as sensitive, invite only.
- Implement access likewise, to view, to contribute and so on.

## Screenshots

N/A

## Alternative Approaches Considered

Did you attempt any other approaches that are not documented in code?

## Review Guide

Notes for the reviewer. How to test this change?

## Checklist before requesting a review

- 📖 Read the FMTM Contributing Guide: <https://github.com/hotosm/fmtm/blob/main/CONTRIBUTING.md>
- 📖 Read the HOT Code of Conduct: <https://docs.hotosm.org/code-of-conduct>
- 👷‍♀️ Create small PRs. In most cases, this will be possible.
- ✅ Provide tests for your changes.
- 📝 Use descriptive commit messages.
- 📗 Update any related documentation and include any relevant screenshots.

## [optional] What gif best describes this PR or how it makes you feel?

spwoodcock commented on src/backend/app/auth/auth_deps.py:
```diff
@@ -268,6 +268,23 @@ async def expire_cookies(response: Response, cookie_names: list[str]) -> Respons
 
 
 ### Endpoint Dependencies ###
+async def public_view(
```
I like this idea, but do we definitely need this extra access level?

If we use `login_required` then the endpoint is truly public and we don't care about auth.

If we use `mapper` role, shouldn't we implement the logic to check for public/private here?

Maybe I haven't thought this through properly

Sujanadh replied:
So implementing login required won't allow users to view the projects without logging in to the FMTM; we need tokens set, the same for the mapper role. So, I would say `login_required` is only public if they have signed up in FMTM.

spwoodcock replied:
Ah of course, my bad there - we do need the new dependency as you have created 👍 
- `login_required` is still needed to add authentication to endpoints that have no associated project, e.g. the helper endpoints.
- I would propose renaming `public_view` to `project_visibility`. The code would read like `Depends(project_visibility)`, which I think is quite intuitive.

spwoodcock replied:
Ah sorry again, I actually looked through your code more thoroughly. I understand what you are trying to do.

We need a dependency to get the currently user, but also fallback to allowing no login!

spwoodcock reviewed (approved):
This is a really nice change - apologies for the rushed review - you documented the change really well in the description😅

spwoodcock commented on src/backend/app/db/models.py:
```diff
@@ -1509,47 +1510,95 @@ async def all(
+            )
+            params["current_user_sub"] = current_user.sub
+
+        if current_user:
```
No need for second `if current_user` here

spwoodcock reviewed (commented):
Happy to merge this as it is. My only comment would be, as the `.all` method has grown so huge, we could probably make functions to make it easier to understand, e.g.

```python
def build_visibility_filter(is_superadmin, managed_org_ids, current_user_sub):
    if is_superadmin:
        return None, {}
    elif managed_org_ids:
        return (
            "(p.visibility = 'PUBLIC' OR p.organisation_id = ANY(%(managed_org_ids)s))",
            {"managed_org_ids": managed_org_ids}
        )
    else:
        return (
            "(p.visibility = 'PUBLIC' OR (p.visibility = 'PRIVATE' AND ur.user_sub = %(current_user_sub)s))",
            {"current_user_sub": current_user_sub}
        )

visibility_filter, visibility_params = build_visibility_filter(...)
if visibility_filter:
    filters.append(visibility_filter)
params.update(visibility_params)
```

spwoodcock commented:
Happy to merge this as it is. My only comment would be, as the `.all` method has grown so huge, we could probably make functions to make it easier to understand, e.g.

```python
def build_visibility_filter(is_superadmin, managed_org_ids, current_user_sub):
    if is_superadmin:
        return None, {}
    elif managed_org_ids:
        return (
            "(p.visibility = 'PUBLIC' OR p.organisation_id = ANY(%(managed_org_ids)s))",
            {"managed_org_ids": managed_org_ids}
        )
    else:
        return (
            "(p.visibility = 'PUBLIC' OR (p.visibility = 'PRIVATE' AND ur.user_sub = %(current_user_sub)s))",
            {"current_user_sub": current_user_sub}
        )

visibility_filter, visibility_params = build_visibility_filter(...)
if visibility_filter:
    filters.append(visibility_filter)
params.update(visibility_params)
```

Sujanadh commented:
## Updates:
- Added a `needs_user_roles_join` parameter to control when the `user_roles` join should be included
- Created separate function to get the user access level `_get_user_access_level`
- Refactored the visibility filter logic into separate function `_build_visibility_filter`
- Separated minimal and full query to separate functions :-  `_minimal_select_query` , `_full_select_query`

spwoodcock reviewed (approved):
This is really nice - great work 👏
